lib/xcodeproj/project/object/native_target.rb

```diff
@@ -407,18 +407,21 @@ def symbol_type
         #
         def add_file_references(file_references, compiler_flags = {})
           file_references.map do |file|
-            build_file = project.new(PBXBuildFile)
-            build_file.file_ref = file
-
             extension = File.extname(file.path).downcase
             header_extensions = Constants::HEADER_FILES_EXTENSIONS
-            if header_extensions.include?(extension)
-              headers_build_phase.files << build_file
-            else
-              if compiler_flags && !compiler_flags.empty?
-                build_file.settings = { 'COMPILER_FLAGS' => compiler_flags }
+            is_header_phase = header_extensions.include?(extension)
+            phase = is_header_phase ? headers_build_phase : source_build_phase
+
+            unless build_file = phase.build_file(file)
+              build_file = project.new(PBXBuildFile)
+              build_file.file_ref = file
+              phase.files << build_file
+            end
+
+            if compiler_flags && !compiler_flags.empty? && !is_header_phase
+              (build_file.settings ||= {}).merge!('COMPILER_FLAGS' => compiler_flags) do |_, old, new|
+                (old || '').+(new || '').split(' ').uniq.join(' ')
```

Comment:
I wonder whether `split(' ')` is enough. Are there any compiler flags allowed on a per file level, which may contain spaces? Something like `-option 'value'` would very likely break. Perhaps we should stick with normal string concatenation.